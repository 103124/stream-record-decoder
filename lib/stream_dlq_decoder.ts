import * as cdk from 'aws-cdk-lib';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as sqs from 'aws-cdk-lib/aws-sqs';
import * as lambdaEventSources from 'aws-cdk-lib/aws-lambda-event-sources';
import { Construct } from 'constructs';
import * as path from 'path';
import {aws_dynamodb, Duration} from "aws-cdk-lib";

export interface StreamDLQDecoderProps {
  /**
   * The Dead Letter Queue (DLQ) the Lambda function reads from
   */
  deadLetterQueue: sqs.IQueue;
  /**
   * The DynamoDB table whose stream the Lambda function reads from
   */
  dynamoTable: aws_dynamodb.ITable;
  /**
   * The destination queue the Lambda function writes to
   */
  destinationQueueProps: sqs.QueueProps;
}

export class StreamDLQDecoder extends Construct {
  public readonly function: lambda.Function;
  public readonly destinationQueue: sqs.IQueue;

  constructor(scope: Construct, id: string, props: StreamDLQDecoderProps) {
    super(scope, id)

    const role = new cdk.aws_iam.Role(this, `${id}Role`, {
      assumedBy: new cdk.aws_iam.ServicePrincipal('lambda.amazonaws.com'),
    })
    props.dynamoTable.grantStreamRead(role)

    this.function = new lambda.Function(this, `${id}Function`, {
      runtime: lambda.Runtime.PROVIDED_AL2023,
      handler: 'bootstrap',
      code: lambda.Code.fromAsset(path.join(__dirname, 'lambda')),
      role: role,
      deadLetterQueueEnabled: false,
    })
    this.function.addEventSource(new lambdaEventSources.SqsEventSource(props.deadLetterQueue))

   const defaultQueueProps: sqs.QueueProps = {
     retentionPeriod: Duration.days(14),
   }
   this.destinationQueue = new sqs.Queue(this, `${id}Queue`, {...defaultQueueProps, ...props.destinationQueueProps})
   this.destinationQueue.grantSendMessages(role)
  }
}