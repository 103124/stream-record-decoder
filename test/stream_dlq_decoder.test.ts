import * as cdk from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { StreamDLQDecoder } from '../lib/stream_dlq_decoder';

test('StreamDLQDecoder creates expected resources', () => {
  const app = new cdk.App();
  const stack = new cdk.Stack(app, 'TestStack');
  const mockDLQ = cdk.aws_sqs.Queue.fromQueueArn(
    stack,
    'MockDLQ',
    'arn:aws:sqs:us-east-1:123456789012:MockDLQ'
  )
  const mockTable = cdk.aws_dynamodb.Table.fromTableAttributes(stack, 'MockTable', {
    tableArn: 'arn:aws:dynamodb:us-east-1:123456789012:table/MockTable',
    tableStreamArn: 'arn:aws:dynamodb:us-east-1:123456789012:table/MockTable/stream/2023-11-22T00:00:00.000',
  })

  new StreamDLQDecoder(stack, 'MyTestConstruct', {
    deadLetterQueue: mockDLQ,
    dynamoTable: mockTable,
    destinationQueueProps: {}
  })

  const template = Template.fromStack(stack);
  expect(template).toMatchSnapshot();
});
