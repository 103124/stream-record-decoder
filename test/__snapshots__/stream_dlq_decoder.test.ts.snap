// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StreamDLQDecoder creates expected resources 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "MyTestConstructMyTestConstructFunction59CBE4CB": {
      "DependsOn": [
        "MyTestConstructMyTestConstructRoleDefaultPolicyC6859E17",
        "MyTestConstructMyTestConstructRole732E1DEC",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "43223157c0b5ad897d02b723fb9f17bb12c9f86bfc078cd6cca6a37ff6c6c402.zip",
        },
        "Handler": "bootstrap",
        "Role": {
          "Fn::GetAtt": [
            "MyTestConstructMyTestConstructRole732E1DEC",
            "Arn",
          ],
        },
        "Runtime": "provided.al2023",
      },
      "Type": "AWS::Lambda::Function",
    },
    "MyTestConstructMyTestConstructFunctionSqsEventSourceTestStackMockDLQ51EA77B3751DAE56": {
      "Properties": {
        "EventSourceArn": "arn:aws:sqs:us-east-1:123456789012:MockDLQ",
        "FunctionName": {
          "Ref": "MyTestConstructMyTestConstructFunction59CBE4CB",
        },
      },
      "Type": "AWS::Lambda::EventSourceMapping",
    },
    "MyTestConstructMyTestConstructQueue2986537C": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "MessageRetentionPeriod": 1209600,
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "MyTestConstructMyTestConstructRole732E1DEC": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "MyTestConstructMyTestConstructRoleDefaultPolicyC6859E17": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "dynamodb:ListStreams",
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": [
                "dynamodb:DescribeStream",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
              ],
              "Effect": "Allow",
              "Resource": "arn:aws:dynamodb:us-east-1:123456789012:table/MockTable/stream/2023-11-22T00:00:00.000",
            },
            {
              "Action": [
                "sqs:ReceiveMessage",
                "sqs:ChangeMessageVisibility",
                "sqs:GetQueueUrl",
                "sqs:DeleteMessage",
                "sqs:GetQueueAttributes",
              ],
              "Effect": "Allow",
              "Resource": "arn:aws:sqs:us-east-1:123456789012:MockDLQ",
            },
            {
              "Action": [
                "sqs:SendMessage",
                "sqs:GetQueueAttributes",
                "sqs:GetQueueUrl",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "MyTestConstructMyTestConstructQueue2986537C",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "MyTestConstructMyTestConstructRoleDefaultPolicyC6859E17",
        "Roles": [
          {
            "Ref": "MyTestConstructMyTestConstructRole732E1DEC",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
